dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/LoopUnrolling.java

```diff
@@ -79,74 +85,134 @@ public void run(Program program) {
+        Preconditions.checkArgument(loopBackJump.getUId() < 0, "The loop has already been unrolled");
+
+        // (1) Collect continue points of the loop
+        List<CondJump> continues = new ArrayList<>();
+        for (Event e = loopBegin; e != null && e != loopBackJump; e = e.getSuccessor()) {
+            if (e instanceof CondJump && ((CondJump)e).getLabel() == loopBegin) {
+                continues.add((CondJump) e);
+            }
+        }
+        continues.add(loopBackJump);
+
+        // (2) Collect forward jumps from the outside into the loop
+        List<CondJump> enterJumps = new ArrayList<>();
+        for (Event e = loopBegin; e != null && e != loopBackJump; e = e.getSuccessor()) {
+            if (e instanceof Label) {
+                Label label = (Label) e;
+                label.getListeners().stream()
+                        .filter(j -> j instanceof CondJump && j.getOId() < loopBegin.getOId())
+                        .map(CondJump.class::cast)
+                        .forEach(enterJumps::add);
+            }
+        }
+
+        int iterCounter = 0;
+        while (--bound >= 0) {
+            iterCounter++;
+            if (bound == 0) {
+                Label exit = (Label) loopBackJump.getThread().getExit();
+                loopBegin.setName(loopBegin.getName() + "_" + iterCounter);
+                for (CondJump cont : continues) {
+                    if (!cont.isGoto()) {
+                        logger.warn("Conditional jump {} was replaced by unconditional bound event", cont);
                     }
-                    predecessor.setSuccessor(newPred);
-                } else {
-                    newPred = predecessor;
+                    CondJump boundEvent = EventFactory.newGoto(exit);
+                    boundEvent.addFilters(cont.getFilters()); // Keep tags of original jump.
+                    boundEvent.addFilters(Tag.BOUND, Tag.NOOPT);
+
+                    cont.getPredecessor().setSuccessor(boundEvent);
+                    boundEvent.setSuccessor(cont.getSuccessor());
+                    cont.delete();
                 }
             } else {
-                newPred = cur;
-                if (predecessor != null) {
-                    // This check must be done inside this if
-                    // Needed for the current implementation of copy in If events
-                    //TODO: Is this needed anymore since we got rid of If events?
-                    if (bound != 1) {
-                        newPred = cur.getCopy();
-                    }
-                    predecessor.setSuccessor(newPred);
+                Map<Event, Event> copyCtx = new HashMap<>();
+                List<Event> copies = copyPath(loopBegin, loopBackJump, copyCtx);
+                ((Label)copyCtx.get(loopBegin)).setName(loopBegin.getName() + "_" + iterCounter);
+                //TODO: The following was testing code and needs to be updated for real usage.
```

Comment:
what about this?

Reply:
It's a matter of whether we want special annotation events or not. This code is also the reason why I didn't create a PR for the branch yet. We need to discuss if we want to have annotation events and what their purpose should be. But we can always add them later if needed, so I'm fine with just getting rid of that code for now.

Reply:
I removed the annotation code.